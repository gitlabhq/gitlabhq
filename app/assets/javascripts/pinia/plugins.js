import { cloneDeep, isEqual } from 'lodash';

/**
 * @typedef {import('pinia').PiniaPluginContext} PiniaPluginContext
 * @typedef {import('vuex').Store} VuexStore
 */

/**
 * @param {PiniaPluginContext} context
 */
// never use this for regular stores, only use this if you have circular dependencies during Vuex migration
export const globalAccessorPlugin = (context) => {
  // eslint-disable-next-line no-param-reassign
  context.store.tryStore = (storeName) => {
    // eslint-disable-next-line no-underscore-dangle
    const anotherStore = context.pinia._s.get(storeName);

    if (!anotherStore) {
      const hookName = `use${storeName.charAt(0).toUpperCase() + storeName.slice(1)}`;
      throw new ReferenceError(
        `Store '${storeName}' doesn't exist. Ensure you have called ${hookName}.`,
      );
    }

    return anotherStore;
  };
};

/**
 * @param {PiniaPluginContext} context
 */
// use this only for component migration
export const syncWithVuex = (context) => {
  let unsubPreviousVuex;
  let unsubPreviousPinia;

  const syncWith = (config) => {
    unsubPreviousVuex?.();
    unsubPreviousPinia?.();
    const {
      store: vuexStore,
      name: vuexName,
      namespaced,
    } = /** @type {{ store: VuexStore, [name]: string, [namespaced]: boolean }} */ config;
    const getVuexState = vuexName ? () => vuexStore.state[vuexName] : () => vuexStore.state;
    if (!isEqual(context.store.$state, getVuexState())) {
      Object.entries(getVuexState()).forEach(([key, value]) => {
        // we can't use store.$patch here because it will merge state, but we need to overwrite it
        // eslint-disable-next-line no-param-reassign
        context.store[key] = cloneDeep(value);
      });
    }

    let committing = false;

    unsubPreviousVuex = vuexStore.subscribe(
      (mutation) => {
        if (committing) return;
        const { payload, type } = mutation;
        const [prefixOrName, mutationName] = type.split('/');
        committing = true;
        if (!mutationName && prefixOrName in context.store) {
          context.store[prefixOrName](cloneDeep(payload));
        } else if (prefixOrName === vuexName && mutationName in context.store) {
          context.store[mutationName](cloneDeep(payload));
        }
        committing = false;
      },
      { prepend: true },
    );

    unsubPreviousPinia = context.store.$onAction(({ name: mutationName, args }) => {
      if (committing) return;
      const fullMutationName = namespaced ? `${vuexName}/${mutationName}` : mutationName;
      // eslint-disable-next-line no-underscore-dangle
      if (!(fullMutationName in vuexStore._mutations)) return;
      committing = true;
      vuexStore.commit(fullMutationName, ...cloneDeep(args));
      committing = false;
    });
  };

  const initialConfig = context.options.syncWith;
  if (initialConfig) syncWith(initialConfig);

  return { syncWith };
};
