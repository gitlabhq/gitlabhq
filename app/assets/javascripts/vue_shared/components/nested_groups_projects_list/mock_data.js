/* eslint-disable @gitlab/require-i18n-strings */
import { uniqueId } from 'lodash';
import { ACTION_EDIT, ACTION_DELETE } from '~/vue_shared/components/list_actions/constants';
import { slugify } from '~/lib/utils/text_utility';
import { LIST_ITEM_TYPE_PROJECT, LIST_ITEM_TYPE_GROUP } from './constants';

const makeGroup = ({ name, fullName, childrenToLoad = [] }) => {
  const fullPath = slugify(fullName);
  const id = parseInt(uniqueId(), 10);

  return {
    type: LIST_ITEM_TYPE_GROUP,
    markedForDeletionOn: null,
    permanentDeletionDate: '2025-02-26',
    fullPath,
    descriptionHtml:
      '<p data-sourcepos="1:1-1:64" dir="auto">Sapiente excepturi est eos corrupti possimus praesentium quidem.</p>',
    avatarUrl: null,
    descendantGroupsCount: 0,
    projectsCount: 1,
    groupMembersCount: 5,
    visibility: 'public',
    createdAt: '2024-09-05T11:04:39Z',
    updatedAt: '2024-10-03T18:09:02Z',
    isLinkedToSubscription: true,
    id,
    avatarLabel: name,
    fullName,
    webUrl: `https://gdk.test:3443/${fullPath}`,
    accessLevel: { integerValue: 50 },
    editPath: `/${fullPath}/edit`,
    availableActions: [ACTION_EDIT, ACTION_DELETE],
    children: [],
    childrenToLoad: childrenToLoad.map((child) => ({ ...child, parent: { id } })),
    hasChildren: childrenToLoad.length > 0,
    childrenCount: childrenToLoad.length,
    childrenLoading: false,
  };
};

const makeProject = ({ name, nameWithNamespace }) => {
  const fullPath = slugify(nameWithNamespace);

  return {
    type: LIST_ITEM_TYPE_PROJECT,
    markedForDeletionOn: null,
    permanentDeletionDate: '2025-02-26',
    fullPath,
    archived: false,
    webUrl: `https://gdk.test:3443/${fullPath}`,
    topics: [],
    forksCount: 1,
    avatarUrl: null,
    starCount: 1,
    visibility: 'public',
    openMergeRequestsCount: 4,
    openIssuesCount: 47,
    descriptionHtml:
      '<p data-sourcepos="1:1-1:503" dir="auto">Qui nostrum occaecati eum quo dicta quam. Qui nostrum occaecati eum quo dicta quam. Qui nostrum occaecati eum quo dicta quam. Qui nostrum occaecati eum quo dicta quam. Qui nostrum occaecati eum quo dicta quam. Qui nostrum occaecati eum quo dicta quam. Qui nostrum occaecati eum quo dicta quam. Qui nostrum occaecati eum quo dicta quam. Qui nostrum occaecati eum quo dicta quam. Qui nostrum occaecati eum quo dicta quam. Qui nostrum occaecati eum quo dicta quam. Qui nostrum occaecati eum quo dicta quam.</p>',
    createdAt: '2024-09-05T11:04:37Z',
    updatedAt: '2024-12-12T18:33:54Z',
    lastActivityAt: '2024-12-12T18:33:54Z',
    userPermissions: {
      removeProject: true,
      viewEditPage: true,
    },
    isCatalogResource: false,
    exploreCatalogPath: null,
    pipeline: {
      detailedStatus: {
        id: 'failed-566-566',
        icon: 'status_failed',
        text: 'Failed',
        detailsPath: `/${fullPath}/-/pipelines/566`,
      },
    },
    id: parseInt(uniqueId(), 10),
    name,
    nameWithNamespace,
    avatarLabel: name,
    mergeRequestsAccessLevel: 'ENABLED',
    issuesAccessLevel: 'ENABLED',
    forkingAccessLevel: 'ENABLED',
    isForked: false,
    accessLevel: { integerValue: 50 },
    availableActions: [ACTION_EDIT, ACTION_DELETE],
    editPath: `/${fullPath}/edit`,
  };
};

export const projectA = makeProject({
  name: 'Project A',
  nameWithNamespace: 'Top-level group A / Subgroup / Nested subgroup / Project A',
});

export const projectB = makeProject({
  name: 'Project B',
  nameWithNamespace: 'Top-level group A / Project B',
});

export const nestedSubgroup = makeGroup({
  name: 'Nested subgroup',
  fullName: 'Top-level group A / Subgroup / Nested subgroup',
  childrenToLoad: [projectA],
});

export const subgroup = makeGroup({
  name: 'Subgroup',
  fullName: 'Top-level group A / Subgroup',
  childrenToLoad: [nestedSubgroup],
});

export const topLevelGroupA = makeGroup({
  name: 'Top-level group A',
  fullName: 'Top-level group A',
  childrenToLoad: [subgroup, projectB],
});

export const topLevelGroupB = makeGroup({
  name: 'Top-level group B',
  fullName: 'Top-level group B',
});

export const items = [topLevelGroupA, topLevelGroupB];
