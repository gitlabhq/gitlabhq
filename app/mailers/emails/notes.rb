# frozen_string_literal: true

module Emails
  module Notes
    def note_commit_email(recipient_id, note_id, reason = nil)
      setup_note_mail(note_id, recipient_id)

      @commit = @note.noteable
      @target_url = project_commit_url(*note_target_url_options)
      mail_answer_note_thread(@commit, @note, note_thread_options(reason))
    end

    def note_issue_email(recipient_id, note_id, reason = nil)
      setup_note_mail(note_id, recipient_id)

      @issue = @note.noteable
      @target_url = Gitlab::UrlBuilder.build(@issue, **note_target_url_query_params)
      mail_answer_note_thread(
        @issue,
        @note,
        note_thread_options(
          reason,
          confidentiality: @issue.confidential?
        )
      )
    end

    def note_merge_request_email(recipient_id, note_id, reason = nil)
      setup_note_mail(note_id, recipient_id)

      @merge_request = @note.noteable
      @target_url = project_merge_request_url(*note_target_url_options)
      mail_answer_note_thread(@merge_request, @note, note_thread_options(reason))
    end

    def note_snippet_email(recipient_id, note_id, reason = nil)
      setup_note_mail(note_id, recipient_id)
      @snippet = @note.noteable

      case @snippet
      when ProjectSnippet
        @target_url = project_snippet_url(*note_target_url_options)
      when Snippet
        @target_url = gitlab_snippet_url(@note.noteable)
      end

      mail_answer_note_thread(@snippet, @note, note_thread_options(reason))
    end

    def note_design_email(recipient_id, note_id, reason = nil)
      setup_note_mail(note_id, recipient_id)

      design = @note.noteable
      @target_url = ::Gitlab::Routing.url_helpers.designs_project_issue_url(
        @note.resource_parent,
        design.issue,
        note_target_url_query_params.merge(vueroute: design.filename)
      )
      mail_answer_note_thread(design, @note, note_thread_options(reason))
    end

    def note_wiki_page_email(recipient_id, note_id, reason = nil)
      setup_note_mail(note_id, recipient_id)

      @wiki_page = @note.noteable
      @target_url = Gitlab::UrlBuilder.build(@wiki_page, **note_target_url_query_params)

      mail_answer_note_thread(@wiki_page, @note, note_thread_options(reason))
    end

    private

    def note_target_url_options
      [@project || @group, @note.noteable, note_target_url_query_params]
    end

    def note_target_url_query_params
      { anchor: "note_#{@note.id}" }
    end

    def note_thread_options(reason, confidentiality: nil)
      {
        from: sender(@note.author_id),
        to: @recipient.notification_email_for(@project&.group || @group),
        subject: subject("#{@note.noteable.title} (#{@note.noteable.reference_link_text})"),
        'X-GitLab-NotificationReason' => reason
      }.tap do |options|
        options['X-GitLab-ConfidentialIssue'] = confidentiality.to_s unless confidentiality.nil?
      end
    end

    def setup_note_mail(note_id, recipient_id)
      # `note_id` is a `Note` when originating in `NotifyPreview`
      @note = note_id.is_a?(Note) ? note_id : Note.find(note_id)
      @project = @note.project
      @group = @note.noteable.try(:group)
      @group ||= @note.noteable.resource_parent if @note.noteable.try(:resource_parent).is_a?(Group)
      @recipient = User.find(recipient_id)

      @sent_notification = SentNotification.record_note(@note, recipient_id) if (@project || @group) && @note.persisted?
    end
  end
end

Emails::Notes.prepend_mod_with('Emails::Notes')
