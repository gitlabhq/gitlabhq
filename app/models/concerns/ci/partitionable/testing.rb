# frozen_string_literal: true

module Ci
  module Partitionable
    module Testing
      InclusionError = Class.new(StandardError)

      PARTITIONABLE_MODELS = %w[
        CommitStatus
        Ci::BuildExecutionConfig
        Ci::BuildMetadata
        Ci::BuildName
        Ci::BuildNeed
        Ci::BuildReportResult
        Ci::BuildRunnerSession
        Ci::BuildSource
        Ci::BuildTag
        Ci::BuildTraceChunk
        Ci::BuildTraceMetadata
        Ci::BuildPendingState
        Ci::DailyBuildGroupReportResult
        Ci::JobAnnotation
        Ci::JobArtifact
        Ci::JobArtifactReport
        Ci::JobInput
        Ci::JobVariable
        Ci::Pipeline
        Ci::PendingBuild
        Ci::RunningBuild
        Ci::RunnerManagerBuild
        Ci::PipelineArtifact
        Ci::PipelineChatData
        Ci::PipelineMessage
        Ci::PipelineMetadata
        Ci::PipelineVariable
        Ci::Sources::Pipeline
        Ci::Sources::Project
        Ci::Stage
        Ci::UnitTestFailure
        Ci::Workloads::Workload
      ].freeze

      def self.check_inclusion(klass)
        return if partitionable_models.include?(klass.name)

        raise Partitionable::Testing::InclusionError,
          "#{klass} must be included in PARTITIONABLE_MODELS"

      rescue InclusionError => e
        Gitlab::ErrorTracking.track_and_raise_for_dev_exception(e)
      end

      def self.partitionable_models
        PARTITIONABLE_MODELS
      end
    end
  end
end

Ci::Partitionable::Testing.prepend_mod
