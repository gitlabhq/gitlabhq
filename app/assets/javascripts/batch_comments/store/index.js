import { defineStore } from 'pinia';
import mrNotes from '~/mr_notes/stores';
import * as actions from './actions';
import mutations from './mutations';
import * as getters from './getters';

export const useBatchComments = defineStore('batchComments', {
  syncWith: {
    store: mrNotes,
    name: 'batchComments',
    namespaced: true,
  },
  state() {
    return {
      isDraftsFetched: false,
      drafts: [],
      isPublishing: false,
      currentlyPublishingDrafts: [],
      shouldAnimateReviewButton: false,
      isMergeRequest: false,
      drawerOpened: false,
      // TODO: this gets populated from the sidebar_reviewers.vue, we should have a separate store for this
      isReviewer: false,
    };
  },
  actions: {
    ...mutations,
    ...actions,
  },
  getters,
});
