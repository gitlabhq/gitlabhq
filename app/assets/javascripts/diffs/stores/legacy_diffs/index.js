import { defineStore } from 'pinia';
import { INLINE_DIFF_VIEW_TYPE } from '~/diffs/constants';
import * as actions from './actions';
import mutations from './mutations';
import * as getters from './getters';

export const useLegacyDiffs = defineStore('legacyDiffs', {
  state() {
    return {
      isLoading: true,
      isTreeLoaded: false,
      batchLoadingState: null,
      retrievingBatches: false,
      addedLines: null,
      removedLines: null,
      endpoint: '',
      endpointBatch: '',
      endpointMetadata: '',
      endpointCoverage: '',
      endpointUpdateUser: '',
      endpointDiffForPath: '',
      perPage: undefined,
      basePath: '',
      commit: null,
      startVersion: null, // Null unless a target diff is selected for comparison that is not the "base" diff
      diffFiles: [],
      coverageFiles: {},
      coverageLoaded: false,
      mergeRequestDiffs: [],
      mergeRequestDiff: null,
      diffViewType: INLINE_DIFF_VIEW_TYPE,
      tree: [],
      treeEntries: {},
      currentDiffFileId: '',
      projectPath: '',
      viewedDiffFileIds: {},
      commentForms: [],
      highlightedRow: null,
      renderTreeList: true,
      showWhitespace: true,
      viewDiffsFileByFile: false,
      fileFinderVisible: false,
      dismissEndpoint: '',
      showSuggestPopover: true,
      defaultSuggestionCommitMessage: '',
      mrReviews: {},
      latestDiff: true,
      virtualScrollerDisabled: false,
      linkedFileHash: null,
    };
  },
  actions: {
    ...mutations,
    ...actions,
  },
  getters,
});
