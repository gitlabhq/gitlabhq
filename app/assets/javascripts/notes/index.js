import Vue from 'vue';
import VueApollo from 'vue-apollo';
import { apolloProvider } from '~/graphql_shared/issuable_client';
import { convertToGraphQLId } from '~/graphql_shared/utils';
import { parseBoolean } from '~/lib/utils/common_utils';
import { getLocationHash } from '~/lib/utils/url_utility';
import { pinia } from '~/pinia/instance';
import { useNotes } from '~/notes/store/legacy_notes';
import NotesApp from './components/notes_app.vue';
import { store } from './stores';
import { getNotesFilterData } from './utils/get_notes_filter_data';

export default ({ editorAiActions = [] } = {}) => {
  const el = document.getElementById('js-vue-notes');
  if (!el) {
    return;
  }

  Vue.use(VueApollo);

  const notesFilterProps = getNotesFilterData(el);
  const showTimelineViewToggle = parseBoolean(el.dataset.showTimelineViewToggle);

  const notesDataset = el.dataset;
  const parsedUserData = JSON.parse(notesDataset.currentUserData);
  const noteableData = JSON.parse(notesDataset.noteableData);
  let currentUserData = {};

  noteableData.noteableType = notesDataset.noteableType;
  noteableData.targetType = notesDataset.targetType;
  noteableData.discussion_locked = parseBoolean(noteableData.discussion_locked);

  if (parsedUserData) {
    currentUserData = {
      id: parsedUserData.id,
      name: parsedUserData.name,
      username: parsedUserData.username,
      avatar_url: parsedUserData.avatar_path || parsedUserData.avatar_url,
      path: parsedUserData.path,
      can_add_timeline_events: parseBoolean(notesDataset.canAddTimelineEvents),
    };
  }

  const notesData = JSON.parse(notesDataset.notesData);

  useNotes().syncWith({ store });

  store.dispatch('setNotesData', notesData);
  store.dispatch('setNoteableData', noteableData);
  store.dispatch('setUserData', currentUserData);
  store.dispatch('setTargetNoteHash', getLocationHash());
  store.dispatch('fetchNotes');

  // eslint-disable-next-line no-new
  new Vue({
    el,
    name: 'NotesRoot',
    components: {
      NotesApp,
    },
    store,
    pinia,
    apolloProvider,
    provide: {
      showTimelineViewToggle,
      reportAbusePath: notesDataset.reportAbusePath,
      newCommentTemplatePaths: JSON.parse(notesDataset.newCommentTemplatePaths),
      resourceGlobalId: convertToGraphQLId(noteableData.noteableType, noteableData.id),
      legacyEditorAiActions: editorAiActions.map((factory) => factory(noteableData)),
      newCustomEmojiPath: notesDataset.newCustomEmojiPath,
    },
    data() {
      return {
        noteableData,
        currentUserData,
        notesData: JSON.parse(notesDataset.notesData),
      };
    },
    render(createElement) {
      return createElement('notes-app', {
        props: {
          noteableData: this.noteableData,
          notesData: this.notesData,
          userData: this.currentUserData,
          ...notesFilterProps,
        },
      });
    },
  });
};
