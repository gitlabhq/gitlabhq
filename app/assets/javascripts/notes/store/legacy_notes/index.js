import { defineStore } from 'pinia';
import { ASC, MR_FILTER_OPTIONS } from '~/notes/constants';
import * as actions from './actions';
import mutations from './mutations';
import * as getters from './getters';

export const useNotes = defineStore('legacyNotes', {
  state() {
    return {
      discussions: [],
      discussionSortOrder: ASC,
      persistSortOrder: true,
      convertedDisscussionIds: [],
      targetNoteHash: null,
      lastFetchedAt: null,
      currentDiscussionId: null,
      batchSuggestionsInfo: [],
      currentlyFetchingDiscussions: false,
      doneFetchingBatchDiscussions: false,
      /**
       * selectedCommentPosition & selectedCommentPositionHover structures are the same as `position.line_range`:
       * {
       *  start: { line_code: string, new_line: number, old_line:number, type: string },
       *  end: { line_code: string, new_line: number, old_line:number, type: string },
       * }
       */
      selectedCommentPosition: null,
      selectedCommentPositionHover: null,

      // View layer
      isToggleStateButtonLoading: false,
      isNotesFetched: false,
      isLoading: true,
      isLoadingDescriptionVersion: false,
      isPromoteCommentToTimelineEventInProgress: false,

      // holds endpoints and permissions provided through haml
      notesData: {
        markdownDocsPath: '',
      },
      userData: {},
      noteableData: {
        discussion_locked: false,
        confidential: false, // TODO: Move data like this to Issue Store, should not be apart of notes.
        current_user: {},
        preview_note_path: 'path/to/preview',
      },
      isResolvingDiscussion: false,
      commentsDisabled: false,
      resolvableDiscussionsCount: 0,
      unresolvedDiscussionsCount: 0,
      descriptionVersions: {},
      isTimelineEnabled: false,
      isFetching: false,
      isPollingInitialized: false,
      mergeRequestFilters: MR_FILTER_OPTIONS.map((f) => f.value),
    };
  },
  actions: {
    ...mutations,
    ...actions,
  },
  getters: {
    ...getters,
  },
});
