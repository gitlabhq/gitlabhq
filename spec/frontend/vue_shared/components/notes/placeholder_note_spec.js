import { shallowMount } from '@vue/test-utils';
import Vue from 'vue';
import { PiniaVuePlugin } from 'pinia';
import { createTestingPinia } from '@pinia/testing';
import IssuePlaceholderNote from '~/vue_shared/components/notes/placeholder_note.vue';
import { userDataMock } from 'jest/notes/mock_data';
import { globalAccessorPlugin } from '~/pinia/plugins';
import { useLegacyDiffs } from '~/diffs/stores/legacy_diffs';
import { useNotes } from '~/notes/store/legacy_notes';

Vue.use(PiniaVuePlugin);

describe('Issue placeholder note component', () => {
  let pinia;
  let wrapper;

  const findNote = () => wrapper.findComponent({ ref: 'note' });

  const createComponent = (isIndividual = false, propsData = {}) => {
    wrapper = shallowMount(IssuePlaceholderNote, {
      pinia,
      propsData: {
        note: {
          body: 'Foo',
          individual_note: isIndividual,
        },
        ...propsData,
      },
    });
  };

  beforeEach(() => {
    pinia = createTestingPinia({ plugins: [globalAccessorPlugin] });
    useLegacyDiffs();
    useNotes().userData = userDataMock;
  });

  it('matches snapshot', () => {
    createComponent();

    expect(wrapper.element).toMatchSnapshot();
  });

  it('does not add "discussion" class to individual notes', () => {
    createComponent(true);

    expect(findNote().classes()).not.toContain('discussion');
  });

  it('adds "discussion" class to non-individual notes', () => {
    createComponent();

    expect(findNote().classes()).toContain('discussion');
  });
});
