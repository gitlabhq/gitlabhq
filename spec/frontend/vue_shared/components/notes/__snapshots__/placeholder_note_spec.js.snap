// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Issue placeholder note component matches snapshot 1`] = `
<timeline-entry-item-stub
  class="being-posted fade-in-half note note-comment note-wrapper"
>
  <div
    class="gl-float-left timeline-avatar"
  >
    <gl-avatar-link-stub
      href="/root"
    >
      <gl-avatar-stub
        alt="Root"
        entityid="0"
        entityname="root"
        shape="circle"
        size="32"
        src="mock_path"
      />
    </gl-avatar-link-stub>
  </div>
  <div
    class="discussion timeline-content"
  >
    <div
      class="note-header"
    >
      <div
        class="note-header-info"
      >
        <a
          href="/root"
        >
          <span
            class="@sm/panel:gl-inline-block gl-font-bold gl-hidden"
          >
            Root
          </span>
          <span
            class="note-headline-light"
          >
            @root
          </span>
        </a>
      </div>
    </div>
    <div
      class="timeline-discussion-body"
    >
      <div
        class="note-body"
      >
        <div
          class="md note-text"
        >
          <p
            dir="auto"
          >
            Foo
          </p>
        </div>
      </div>
    </div>
  </div>
</timeline-entry-item-stub>
`;
